### PR TITLE
Refactor types in CollectionPicker

### PR DESCRIPTION
**Pull request by kamilmielnik**
Needed for #40298

### Description

`EntityPicker` contains a search component which has been build to support all types of models (see [`SearchModelType`](https://github.com/metabase/metabase/blob/a301cac92a90e3959b38ab2ece328ed7e03be0ad/frontend/src/metabase-types/api/search.ts#L10-L22)).
However - depending on the context - it will usually be used  only with a subset of models, for example:
- in [Collection Picker](https://github.com/metabase/metabase/pull/38315) it is used to select Collections only
- in [Notebook Data Source Picker](https://www.figma.com/file/ff1z9vhHNtgu1LAPhaa5eF/Entity-Picker-Component?type=design&node-id=90-6547&mode=design&t=Ct5ffxWJRLPQc77g-4) it will be used to select Questions, Models, and Tables only

Up until now `EntityPicker` type declarations were tightly coupled with Collections.
This PR removes this coupling and makes the component generic so that we can have type safety in various context without type casting. It's mostly about moving files and extending the types, but I'm sneaking in a few small code changes as well.

Short summary of changes:
- Moves all Collection-related things from `metabase/common/components/EntityPicker` to `metabase/common/components/CollectionPicker`
- Makes almost everything in `metabase/common/components/EntityPicker/` generic, so that it can be used in various contexts
- Removes all casts except 1 in `EntityPickerSearch` and 1 in a unit test
- Removes all `any`s
- Unhookifies `searchFilter` function (it didn't have to be a hook)
- Removes `NestedItemPicker` storybook page - it didn't work and was using an outdated interface of `NestedItemPicker`. Maybe it used to deserve a storybook page at the time of creation, but I don't think it does anymore as its UI part now lives in `ItemList`.

### How to verify

- CI is green
- There are no type casts in `metabase/common/components/CollectionPicker`
- The interface of `CollectionPicker` is Collections-specific

### CODE REVIEW

File: frontend/src/metabase/common/components/EntityPicker/components/CollectionPicker/CollectionPickerModal.tsx [hunk 1]
```diff
@@ -26,6 +24,12 @@ const canSelectItem = (item: CollectionPickerItem | null): boolean => {
   return !!item && item?.can_write !== false;
 };
 
+const searchFilter = (
```
Comment (kamilmielnik) [inline frontend/src/metabase/common/components/EntityPicker/components/CollectionPicker/CollectionPickerModal.tsx:27]: _This change is not type-related._

This function does not need to be a hook.

File: frontend/src/metabase-types/api/search.ts [hunk 2]
```diff
@@ -40,10 +40,23 @@ export interface SearchScore {
...
+}
+
+export interface SearchResults<
+  Id extends SearchResultId = SearchResultId,
+  Model extends SearchModelType = SearchModelType,
+  Result extends BaseSearchResult<Id, Model> = SearchResult<Id, Model>,
+> {
+  data: Result[];
+  models: Model[] | null;
+  available_models: Model[];
```
Comment (ranquild) [inline frontend/src/metabase-types/api/search.ts:59]: I think that's not correct. `available_models` would get the list of all models that satisfy the search query, even if they were not set in `models`.

Comment (kamilmielnik) [reply frontend/src/metabase-types/api/search.ts]: Absolutely, thanks!
9586424f016c294e2007465c4ee21e7ddd886afc

**Review by ranquild — APPROVED** [review]

**Review by iethree — APPROVED** [review]
Very nice work 👏 Thanks! As you can see, we struggled a bit with the types 🥴 

I think that the complexity of these types is actually a smell that the diversity of API responses we get for the same things from the backend could use some work. I'm hopeful that the [openAPI](https://metaboat.slack.com/archives/C010L1Z4F9S/p1710955702931629) docs the backend has recently put in, along with some of the [code gen features of RTK Query](https://redux-toolkit.js.org/rtk-query/usage/code-generation) can help us simplify this in the future.

**Pull request merged**